### PR TITLE
Adjoint canonicalization for CustomOp and MultiRZOp

### PR DESCRIPTION
**Pull request by rmoyard**
**Context:**
We lack canonicalization when the adjoint flag is on.

**Description of the Change:**
- Add canonicalization patterns for cutsomOp (rotations and hermitian)
- Add the patterns in the merge rotation pass.

**Benefits:**
Merge rotations works with adjoint.

### CODE REVIEW

**Comment by rmoyard** [comment]
[sc-75964]

File: mlir/lib/Quantum/IR/QuantumOps.cpp [hunk 1]
```diff
@@ -34,6 +36,50 @@ using namespace catalyst::quantum;
...
+            }
+
+            rewriter.replaceOpWithNewOp<CustomOp>(
+                op, op.getOutQubits().getTypes(), op.getOutCtrlQubits().getTypes(), paramsNeg,
+                op.getInQubits(), name, nullptr, op.getInCtrlQubits(), op.getInCtrlValues());
+
+            return success();
+        }
+        return failure();
+    };
```
Comment (paul0403) [inline mlir/lib/Quantum/IR/QuantumOps.cpp:66]:
```suggestion
    }
```

File: mlir/test/Quantum/CanonicalizationTest.mlir [hunk 2]
```diff
@@ -118,3 +118,43 @@ func.func @test_insert_canonicalize(%r1: !quantum.reg, %i: i64) -> !quantum.bit
...
+    // CHECK: [[reg:%.+]] = quantum.alloc( 2) : !quantum.reg
+    // CHECK: [[qubit1:%.+]] = quantum.extract [[reg]][ 0] : !quantum.reg -> !quantum.bit
+    // CHECK: [[qubit2:%.+]] = quantum.extract [[reg]][ 1] : !quantum.reg -> !quantum.bit
+    %0 = quantum.alloc( 2) : !quantum.reg
+    %1 = quantum.extract %0[ 0] : !quantum.reg -> !quantum.bit
+    %2 = quantum.extract %0[ 1] : !quantum.reg -> !quantum.bit
+
+    // CHECK: [[arg0neg:%.+]] = arith.negf %arg0 : f64
+    // CHECK: [[ret:%.+]]:2 = quantum.multirz([[arg0neg]]) [[qubit1]], [[qubit2]] : !quantum.bit, !quantum.bit
+    // CHECK-NOT: quantum.multirz
```
Comment (paul0403) [inline mlir/test/Quantum/CanonicalizationTest.mlir:156]: Is this `CHECK-NOT` necessary?

Comment (paul0403) [reply mlir/test/Quantum/CanonicalizationTest.mlir:156]: Or did you mean `CHECK-NOT: adjoint` ?

Comment (rmoyard) [reply mlir/test/Quantum/CanonicalizationTest.mlir:156]: Not necessary it is a copy paste

Comment (rmoyard) [inline mlir/test/Quantum/CanonicalizationTest.mlir:156]:
```suggestion
```

Comment (rmoyard) [reply mlir/test/Quantum/CanonicalizationTest.mlir]: Removed

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -211,6 +212,9 @@
 
 <h3>Improvements</h3>
 
+* Adjoint canonicalization is now available in MLIR for `CustomOp` and `MultiRZOp`. It can be used
+  with the `--canonicalize` pass in `quantum-opt`.
```
Comment (paul0403) [inline doc/releases/changelog-dev.md:216]:
```suggestion
  with the `--canonicalize` pass in `quantum-opt`.
  [(#1205)](https://github.com/PennyLaneAI/catalyst/pull/1205)
```

**Review by paul0403 — APPROVED** [review]
Always nice to catch bugs before release! :partying_face:

**Pull request merged**